### PR TITLE
Feature/token manager refactoring

### PR DESCRIPTION
**Pull request by Synthezator**

### CODE REVIEW

File: Softeq.XToolkit.DefaultAuthorization/SecuredHttpServiceGate.cs [hunk 1]
```diff
@@ -107,39 +83,51 @@ protected virtual void HandleInvalidResponse(HttpResponse response)
...
 
-        private async Task<ExecutionStatus?> EnsureNoSessionRetrievalIsRunningAsync()
+        private async Task RefreshTokenAsync()
         {
-            if (!_sessionRetrievalDeferral.IsInProgress)
+            ExecutionStatus refreshingTokenResult;
+            if (_refreshingTokenTask == null)
             {
-                return null;
+                _refreshingTokenTask = _sessionApiService.RefreshTokenAsync();
```
Comment (wcoder) [inline Softeq.XToolkit.DefaultAuthorization/SecuredHttpServiceGate.cs:114]: Please revert `deferral` back
The current solution for multiple parallel requests will be trying refresh token for each request.

File: Softeq.XToolkit.DefaultAuthorization/Softeq.XToolkit.DefaultAuthorization.csproj [hunk 2]
```diff
@@ -6,4 +6,7 @@
     <ProjectReference Include="..\Softeq.XToolkit.HttpClient\Softeq.XToolkit.HttpClient.csproj" />
     <ProjectReference Include="..\Softeq.XToolkit.ProjectInfrastructure\Softeq.XToolkit.CrossCutting.csproj" />
   </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="sameerIOTApps.Plugin.SecureStorage" Version="2.5.0" />
+  </ItemGroup>
```
Comment (Synthezator) [inline Softeq.XToolkit.DefaultAuthorization/Softeq.XToolkit.DefaultAuthorization.csproj:11]: Check if we can delete this plugin from platform-specific projects

File: Softeq.XToolkit.DefaultAuthorization/SecuredTokenManager.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        protected SecuredTokenManagerBase()
+        {
+            Token = CrossSecureStorage.Current.GetValue(TokenKey);
+            _tokenExpirationTime = StringToDate(CrossSecureStorage.Current.GetValue(TokenExpirationKey));
+            RefreshToken = CrossSecureStorage.Current.GetValue(RefreshTokenKey);
+        }
+
+        public string Token { get; private set; }
+
+        public bool IsTokenExpired => !_tokenExpirationTime.HasValue ||
```
Comment (wcoder) [inline Softeq.XToolkit.DefaultAuthorization/SecuredTokenManager.cs:27]: I think we should ignore case when `tokenExpirationTimespanInSeconds = 0` (unused by auth provider) and return false

**Review by wcoder — CHANGES_REQUESTED** [review]

File: Softeq.XToolkit.DefaultAuthorization/SecuredTokenManager.cs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            {
+                throw new Exception(SavingResultExceptionMessage);
+            }
+
+            Token = token;
+            _tokenExpirationTime = tokenExpirationTime;
+            RefreshToken = refreshToken;
+        }
+
+        private static string DateToString(DateTime dateTime)
```
Comment (Synthezator) [inline Softeq.XToolkit.DefaultAuthorization/SecuredTokenManager.cs:64]: Create unit-test: date == StringToDate(DateToString(date))

**Review by wcoder — APPROVED** [review]

**Review by wcoder — APPROVED** [review]

**Pull request merged**